src/target/riscv/riscv-013.c

```diff
@@ -3011,34 +3022,39 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
 			}
 			if (size > 1)
 				value |= ((uint32_t) p[1]) << 8;
-			dmi_write(target, DMI_SBDATA0, value);
+			riscv_batch_add_dmi_write(batch, DMI_SBDATA0, value);
 
 			log_memory_access(address + i * size, value, size, false);
 
-			if (info->bus_master_write_delay) {
-				jtag_add_runtest(info->bus_master_write_delay, TAP_IDLE);
-				if (jtag_execute_queue() != ERROR_OK) {
-					LOG_ERROR("Failed to scan idle sequence");
-					return ERROR_FAIL;
-				}
-			}
+			if (riscv_batch_full(batch))
+				break;
 		}
 
+		result = batch_run(target, batch);
+		riscv_batch_free(batch);
+		if (result != ERROR_OK)
+			return result;
+
 		if (read_sbcs_nonbusy(target, &sbcs) != ERROR_OK)
 			return ERROR_FAIL;
 
+		next_address = sb_read_address(target);
```

Comment:
This read and the following check was necessary to catch a system that would potentially reset the debug module in response to a memory write (to a MMIO GPIO device).  Without this check, OpenOCD would hang indefinitely when the debug module was unexpectedly reset.  Hopefully with the specification clarification (riscv/riscv-debug-spec#488) I can convince others that the platform with this behavior should be changed.

In any case, if you feel it is important, I can eliminate this now from the pull request.

Reply:
Why exactly did it end up looping infinitely?
Can you move this check outside of this loop, so it happens once per call to this function instead of once for every 32 words? And you can probably tighten up the check too. sbaddress should contain exactly what you expect. Anything else indicates some kind of failure.

Reply:
If the debug module was spontaneously reset while the memory write was on going, the next_address read back would be zero.  Since zero is less than end_address, it would never exit that loop, but would also never write any bytes (probably due to overflowing the unsigned loop index i).

The check cannot be moved outside the while loop, since it the point of it was to check if it was necessary to terminate the loop early.

I have a pushed an updated branch that I think will address these issues.  next_address is only read back when necessary (when the target is busy), and is only checked when read.  Otherwise, next_address is updated by the loop without reading from the hardware.  This is tested working as before.